diffrax/_integrate.py

```diff
@@ -73,23 +73,57 @@ class State(eqx.Module):
     dense_save_index: Optional[IntScalarLike]
 
 
-def _is_none(x):
+class _ShapeDtypeTypeCoercer(AbstractArray):
+    def __new__(cls, term_type: Any):
+        if not issubclass(term_type, AbstractArray):
+            return term_type
+
+        cls.array_type = jax.ShapeDtypeStruct
+        cls.dtypes = term_type.dtypes
+        cls.dims = term_type.dims
+        cls.index_variadic = term_type.index_variadic
+        cls.dim_str = term_type.dim_str
+        return cls
+
+
+def _is_none(x: Any) -> bool:
     return x is None
 
 
-def _term_compatible(terms, term_structure):
+def _term_compatible(
+    y: PyTree[ArrayLike],
+    args: PyTree[Any],
+    terms: PyTree[AbstractTerm],
+    term_structure: PyTree,
+) -> bool:
     def _check(term_cls, term):
         if get_origin(term_cls) is MultiTerm:
             if isinstance(term, MultiTerm):
                 [_tmp] = get_args(term_cls)
                 assert get_origin(_tmp) in (tuple, Tuple), "Malformed term_structure"
-                if not _term_compatible(term.terms, get_args(_tmp)):
-                    raise ValueError
-            else:
-                raise ValueError
+                if _term_compatible(y, args, term.terms, get_args(_tmp)):
+                    return
         else:
-            if not isinstance(term, term_cls):
-                raise ValueError
+            term_args = get_args(term_cls)
+            n_term_args = len(term_args)
+            if n_term_args >= 1:
+                _term_cls = get_origin(term_cls)
+                vector_field = jax.eval_shape(term.vf, 0.0, y, args)
+                compatible_vf_type = _ShapeDtypeTypeCoercer(term_args[0])
+                vector_field_compatible = isinstance(vector_field, compatible_vf_type)
+                if n_term_args == 2 and vector_field_compatible:
+                    control = jax.eval_shape(term.contr, 0.0, 0.0)
+                    compatible_control_type = _ShapeDtypeTypeCoercer(term_args[1])
+                    control_compatible = isinstance(control, compatible_control_type)
```

Comment:
If the annotation is a more general type that doesn ot support `isinstance`, e.g. `tuple[int, Float[Array, "foo"]]`, then this will fail.

This calls for a runtime type checker!

As it happens, via jaxtyping we already have a transitive dependency on `typeguard==2.13.3` (this exact version), so we can assume it is present. I think add it to `pyproject.toml` to make things explicit, and then you should be able to do:
```python
def better_isinstance(x, annotation):
    @typeguard.typechecked
    def f(y: annotation):
        pass
    try:
        f(x)
    except TypeError:
        return False
    else:
        return True
```